### PR TITLE
Make association source and target reliable in CDB

### PR DESCRIPTION
**Pull request by scosta-obeo**
### #2913 Merge two commands to avoid refresh between them

This commit merge two command for navigability state of association in
CDB to avoid diagram refresh (and unwanted behavior) between them.

### #2913 Make association source and target reliable in CDB

This commit make association source and target reliable in CDB always
have the same source and target and to prevent a change in property from
changing the source or the target.

This commit also add migration to reverse the association edge in the
wrong direction in old project.

### #2913 Migrate all data tests

### CODE REVIEW

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+    if (dEdge.getTarget() instanceof Association association) {
+      Property sourceProp = InformationServices.getService().getAssociationSource(association);
+      Property targetProp = InformationServices.getService().getAssociationTarget(association);
+      AbstractType expectedSource = sourceProp.getAbstractType();
+      AbstractType expectedTarget = targetProp.getAbstractType();
+      if (dEdge.getSourceNode() instanceof DSemanticDecorator sourceNode && dEdge.getTargetNode() instanceof DSemanticDecorator targetNode) {
+        EObject sourceObject = sourceNode.getTarget();
+        EObject targetObject = targetNode.getTarget();
+        return sourceObject == expectedTarget && targetObject == expectedSource;
+      } else {
```
Comment (lredor) [inline core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:318]: A simple "return false;" for all other cases is enough.

Comment (lredor) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:318]: Proposition:
```
if (dEdge.getTarget() instanceof Association association) {
      ...
      if (dEdge.getSourceNode() instanceof DSemanticDecorator sourceNode && dEdge.getTargetNode() instanceof DSemanticDecorator targetNode) {
        ...
        return sourceObject == expectedTarget && targetObject == expectedSource;
      }
}
// not association or wrong conditions
return false;
```

**Review by lredor — CHANGES_REQUESTED** [review]

File: core/plugins/org.polarsys.capella.core.data.information.properties/src/org/polarsys/capella/core/data/information/properties/fields/NavigableCheckbox.java [hunk 2]
```diff
@@ -82,8 +84,10 @@ public void widgetSelected(SelectionEvent event) {
         if (oppositeMember != null) {
           EObject oppositeTypeElement = (EObject) oppositeMember.eGet(ModellingcorePackage.Literals.ABSTRACT_TYPED_ELEMENT__ABSTRACT_TYPE);
           if (oppositeTypeElement != null) {
-            moveDataValue(semanticElement, oppositeTypeElement, CapellacorePackage.Literals.CLASSIFIER__OWNED_FEATURES);
-            addDataValue(ownerElement, InformationPackage.Literals.ASSOCIATION__NAVIGABLE_MEMBERS, semanticElement);
+            AbstractCompoundCommand command = new AbstractCompoundCommand() {};
+            command.append(createMoveDataValueCommand(semanticElement, oppositeTypeElement, CapellacorePackage.Literals.CLASSIFIER__OWNED_FEATURES));
```
[the remark this answers is not in the document]
Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.data.information.properties/src/org/polarsys/capella/core/data/information/properties/fields/NavigableCheckbox.java:88]: Ok, I've added Optional. Note that `createMoveDataValueCommand` does not return null but I changed it to be consistent.

File: core/plugins/org.polarsys.capella.core.sirius.analysis/description/common.odesign
[the remark this answers is not in the document]
Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.sirius.analysis/description/common.odesign:1]: Yes, it's the fix (in part): if I change the direction of the edge, I have to change the arrow.
For example an edge before the fix:
A -------> B, with:
* source=B
* target=A
* sourceArrow=Arrow
* targetArrow=NoDecoration
which becomes:
A -------> B, with:
* source=**A**
* target=**B** (A and B swapped)
needs to work properly:
* sourceArrow=**NoDecoration**
* targetArrow=**Arrow** (sourceArrow and targetArrow swapped in odesign)

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java (hunk 1 above)
Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:318]: I'm not sure it's easier to read (at least not for me). This "return" fallback principle gives the impression at first sight that `false` is always returned, and confuses the cases where `false` is actually returned. For example, it gets harder to read what's going on if `dEdge.getTarget() instanceof Association association` is `true` and `dEdge.getSourceNode() instanceof DSemanticDecorator sourceNode && dEdge.getTargetNode() instanceof DSemanticDecorator targetNode` is `false`.

Comment (lredor) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:318]: With a "more strict" Checkstyle configuration, like in Sirius for example, you can not have three "return" statements.

File: core/plugins/org.polarsys.capella.core.sirius.analysis/description/common.odesign
[the remark this answers is not in the document]
Comment (lredor) [reply core/plugins/org.polarsys.capella.core.sirius.analysis/description/common.odesign:1]: OK, I understand. Maybe a little comment about this in the commit message would be interesting. Proposition: `To have persistent graphical render, the VSM has been changed on some source/target arrows style.`

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java (hunk 1 above)
Comment (lredor) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:318]: But OK with your version (authorized in Capella code style).

File: core/plugins/org.polarsys.capella.core.sirius.analysis/description/common.odesign
[the remark this answers is not in the document]
Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.sirius.analysis/description/common.odesign:1]: Done

**Review by lredor — APPROVED** [review]

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    dEdge.getOwnedStyle().setSourceArrow(targetArrow);
+    dEdge.getOwnedStyle().setTargetArrow(sourceArrow);
+
+    // inverse the custom features sourceArrow and targetArrow
+    String sourceArrowFeatureName = DiagramPackage.eINSTANCE.getEdgeStyle_SourceArrow().getName();
+    String targetArrowFeatureName = DiagramPackage.eINSTANCE.getEdgeStyle_TargetArrow().getName();
+    if (customFeatures.contains(sourceArrowFeatureName) && !customFeatures.contains(targetArrowFeatureName)) {
+      customFeatures.add(targetArrowFeatureName);
+      customFeatures.remove(sourceArrowFeatureName);
+    } else if (!customFeatures.contains(sourceArrowFeatureName) && customFeatures.contains(targetArrowFeatureName)) {
```
Comment (zgrtls) [inline core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:152]: If we go there, we are going to compute again the contains for both. computing it once and using local variables would gain a little computation time

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+      Property targetProp = InformationServices.getService().getAssociationTarget(association);
+      AbstractType expectedSource = sourceProp.getAbstractType();
+      AbstractType expectedTarget = targetProp.getAbstractType();
+      if (dEdge.getSourceNode() instanceof DSemanticDecorator sourceNode
+          && dEdge.getTargetNode() instanceof DSemanticDecorator targetNode) {
+        EObject sourceObject = sourceNode.getTarget();
+        EObject targetObject = targetNode.getTarget();
+        return sourceObject == expectedTarget && targetObject == expectedSource;
+      } else {
+        // error
```
Comment (zgrtls) [inline core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:319]: Maybe add something to some logs or throw an exception if the data are corrupted ?
Silently ignoring errors during migration is the best way to get strange bugs and data corruption visible months later

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java
Comment (zgrtls) [inline core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:1]: I may be missing something but where do you check that the migration is not run twice on the same iard, for example when migrating to 7.0.1 and then again when 7.0.5 is out and the file is migrated again ?

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java (hunk 3 above)
Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java]: done

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java (hunk 4 above)
Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:323]: I added a log message.

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java
Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:1]: There is no marker for migration in Capella. In the context mentioned, the migration will indeed be launched twice. But the check carried out in the `needMigration(DEdge dEdge)` method will therefore allow no changes during the second migration.

Comment (zgrtls) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:1]: That's very resource (and time) expensive for the user if we no longer create old style DEdge objects. Would it be possible to bypass the whole migration by looking, for example, at the version of one of the meta models we know have increase in version 7.0.1  ?

**Comment by lredor** [comment]
> With the lastest commit, there are still 11 failed tests compare to the single one on the master branch. They should be looked at before validating the PR.

Yes, after analysis, the problem is that "DT_Association" has not the same source and target mappings list. With the new solution based on the ID of the source/target to determine with a deterministic way the source and the target, these lists must be the same. @scosta-obeo will push a new PR with a modification of the common.odesign.

File: core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java
Comment (lredor) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:1]: Is there some other migrations with this kind of check ? The problem seems global for all migration. No ?
For example, in Sirius, there is a [common way](https://github.com/eclipse-sirius/sirius-desktop/blob/master/plugins/org.eclipse.sirius/src/org/eclipse/sirius/business/internal/migration/AbstractSiriusMigrationService.java#L274) to address this problematic.

Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:1]: Yes, I'm agree but it's outside the scope of this issue. I suggest you open a new ticket.

Comment (scosta-obeo) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:1]: Yes, I'm agree but it's outside the scope of this issue. I suggest you open a new ticket.

Comment (zgrtls) [reply core/plugins/org.polarsys.capella.core.data.migration/src/org/polarsys/capella/core/data/migration/aird/AssociationCDBMigrationContributor.java:1]: This is a new migration so it is in the scope of this ticket for this migration.
The refactoring to do it correctly in the other migrations will be new tickets.

**Comment by lredor** [comment]
A new test is failing. The problem seems linked to the bump to new "Sirius version", indeed it appears in the last version of the PR where a rebase on master was done.
Analysis is in progress for `org.polarsys.capella.test.migration.ju.testcases.basic.SysmodelMigrationLayout`

**Comment by lredor** [comment]
> > A new test is failing. The problem seems linked to the bump to new "Sirius version", indeed it appears in the last version of the PR where a rebase on master was done. Analysis is in progress for `org.polarsys.capella.test.migration.ju.testcases.basic.SysmodelMigrationLayout`
> 
> @zgrtls The new failing test is also KO on [master branch](https://ci.eclipse.org/capella/job/capella-product/job/master/lastCompletedBuild/testReport/org.polarsys.capella.test.migration.ju.testcases.basic/SysmodelMigrationLayout/SysmodelMigrationLayout/). So the problem is not linked to this PR but probably caused by one of the recent commits (maybe "[releng] Switch to Sirius 7.4.5-S20241126-110306"?). Do you have any idea? It's difficult to see, because I didn't see any trace of the failing tests on the previous PR (for example for #2917).

The problem seems to be caused by the commit ["Fix broken links in doc towads wiki.eclipse.org"](https://github.com/eclipse-capella/capella/commit/e6b28d25b4c547ac1adf5aae8dcfd807ea995f93). Even though I don't see any link, the test is [OK before](https://ci.eclipse.org/capella/job/capella-product/job/master/558/testReport/junit/org.polarsys.capella.test.migration.ju.testcases.basic/SysmodelMigrationLayout/SysmodelMigrationLayout/) in job 558 and [KO after](https://ci.eclipse.org/capella/job/capella-product/job/master/559/testReport/junit/org.polarsys.capella.test.migration.ju.testcases.basic/SysmodelMigrationLayout/SysmodelMigrationLayout/) in job 559. Is there something changed in the infra that can explain the problem?

**Comment by lredor** [comment]
> > > A new test is failing. The problem seems linked to the bump to new "Sirius version", indeed it appears in the last version of the PR where a rebase on master was done. Analysis is in progress for `org.polarsys.capella.test.migration.ju.testcases.basic.SysmodelMigrationLayout`
> > 
> > 
> > @zgrtls The new failing test is also KO on [master branch](https://ci.eclipse.org/capella/job/capella-product/job/master/lastCompletedBuild/testReport/org.polarsys.capella.test.migration.ju.testcases.basic/SysmodelMigrationLayout/SysmodelMigrationLayout/). So the problem is not linked to this PR but probably caused by one of the recent commits (maybe "[releng] Switch to Sirius 7.4.5-S20241126-110306"?). Do you have any idea? It's difficult to see, because I didn't see any trace of the failing tests on the previous PR (for example for #2917).
> 
> The problem seems to be caused by the commit ["Fix broken links in doc towads wiki.eclipse.org"](https://github.com/eclipse-capella/capella/commit/e6b28d25b4c547ac1adf5aae8dcfd807ea995f93). Even though I don't see any link, the test is [OK before](https://ci.eclipse.org/capella/job/capella-product/job/master/558/testReport/junit/org.polarsys.capella.test.migration.ju.testcases.basic/SysmodelMigrationLayout/SysmodelMigrationLayout/) in job 558 and [KO after](https://ci.eclipse.org/capella/job/capella-product/job/master/559/testReport/junit/org.polarsys.capella.test.migration.ju.testcases.basic/SysmodelMigrationLayout/SysmodelMigrationLayout/) in job 559. Is there something changed in the infra that can explain the problem?

I have created this [Gitlab issue](https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/5382) from the discovered problem with @etraisnel2 .

**Review by zgrtls — CHANGES_REQUESTED** [review]
Looks better. We still have 1 failed test : org.polarsys.capella.test.migration.ju.testcases.basic.RelationStability.test

**Pull request merged**